ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c

```diff
@@ -1138,8 +1205,170 @@ typedef struct
   int width;
   int height;
   gfloat prob;
+
+  int tracking_id;
 } detectedObject;
 
+/**
+ * @brief Update centroids with given bounding boxes.
+ */
+static void
+update_centroids (void **pdata, GArray * boxes)
+{
+  guint i, j;
+  bounding_boxes *bdata = *pdata;
+  GArray *centroids = bdata->centroids;
+  GArray *distanceArray = bdata->distanceArray;
+
+  if (boxes->len > bdata->max_centroids_num) {
+    nns_logw ("update_centroids: too many detected objects");
+    return;
+  }
+  /* remove disappeared centroids */
+  i = 0;
+  while (i < centroids->len) {
+    centroid *c = &g_array_index (centroids, centroid, i);
+    if (c->consecutive_disappeared_frames >= bdata->consecutive_disappear_threshold) {
+      g_array_remove_index (centroids, i);
+    } else {
+      i++;
+    }
+  }
+
+  if (centroids->len > bdata->max_centroids_num) {
+    nns_logw ("update_centroids: too many detected centroids");
+    return;
+  }
+  /* if boxes is empty */
+  if (boxes->len == 0U) {
+    guint i;
+    for (i = 0; i < centroids->len; i++) {
+      centroid *c = &g_array_index (centroids, centroid, i);
+
+      if (c->id > 0)
+        c->consecutive_disappeared_frames++;
+    }
+
+    return;
+  }
+  /* initialize centroids with given boxes */
+  if (centroids->len == 0U) {
+    guint i;
+    for (i = 0; i < boxes->len; i++) {
+      detectedObject *box = &g_array_index (boxes, detectedObject, i);
+      centroid c;
+
+      bdata->centroids_last_id++;
+      c.id = bdata->centroids_last_id;
+      c.consecutive_disappeared_frames = 0;
+      c.cx = box->x + box->width / 2;
+      c.cy = box->y + box->height / 2;
+      c.matched_box_idx = i;
+
+      g_array_append_val (centroids, c);
+
+      box->tracking_id = c.id;
+    }
+
+    return;
+  }
+  /* calculate the distance among centroids and boxes */
+  g_array_set_size (distanceArray, centroids->len * boxes->len);
+
+  for (i = 0; i < centroids->len; i++) {
+    centroid *c = &g_array_index (centroids, centroid, i);
+    c->matched_box_idx = 987654321U;
+
+    for (j = 0; j < boxes->len; j++) {
+      detectedObject *box = &g_array_index (boxes, detectedObject, j);
+      distanceArrayData *d =
+          &g_array_index (distanceArray, distanceArrayData,
+          i * centroids->len + j);
+
+      d->centroid_idx = i;
+      d->box_idx = j;
+
+      /* invalid centroid */
+      if (c->id == 0) {
+        d->distance = G_MAXUINT64;
+      } else {
+        /* calculate euclidean distance */
+        d->distance =
+            (c->cx - box->x) * (c->cx - box->x) + (c->cy - box->y) * (c->cy -
+            box->y);
+      }
+    }
+  }
+
+  g_array_sort (distanceArray, distance_compare);
+
+  {
```

Comment:
Is the '{' intended?

Reply:
Yes it is. To avoid compiler's complaint about declaring variables..